File: tunnel-common/src/main/java/com/tuuzed/tunnel/common/interceptor/HttpRequestInterceptor.java
package com.tuuzed.tunnel.common.interceptor;

import com.tuuzed.tunnel.common.proto.ProtoRequest;
import io.netty.handler.codec.http.HttpRequest;
import io.netty.handler.codec.http.HttpResponse;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.net.SocketAddress;

public interface HttpRequestInterceptor {

    HttpRequestInterceptor DEFAULT = new HttpRequestInterceptor() {
        @Nullable
        @Override
        public HttpResponse handleHttpRequest(@NotNull SocketAddress localAddress, @NotNull SocketAddress remoteAddress, @NotNull ProtoRequest protoRequest, @NotNull HttpRequest httpRequest) {
            return null;
        }
    };

    @Nullable
    HttpResponse handleHttpRequest(
        @NotNull SocketAddress localAddress,
        @NotNull SocketAddress remoteAddress,
        @NotNull ProtoRequest protoRequest,
        @NotNull HttpRequest httpRequest
    );

}

File: tunnel2-server/src/main/kotlin/tunnel2/server/interceptor/HttpRequestInterceptor.kt
package tunnel2.server.interceptor

import io.netty.handler.codec.http.HttpRequest
import io.netty.handler.codec.http.HttpResponse
import tunnel2.common.TunnelRequest

import java.net.SocketAddress

interface HttpRequestInterceptor {

    companion object {
        val EMPTY_IMPL = object : HttpRequestInterceptor {}
    }

    fun handleHttpRequest(
        localAddress: SocketAddress,
        remoteAddress: SocketAddress,
        tunnelRequest: TunnelRequest,
        httpRequest: HttpRequest
    ): HttpResponse? = null

}
